# 第一讲 前言

***

<span id="mulu"><font face="黑体" size=6>本章目录</font></span>
* [1 第一讲 前言](#1)
    
    * [1.1 Python 的前世今生](#1.1)
        * [1.1.1 Python的热度](#1.1.1)
        * [1.1.2 为什么 Python 会继续获得程序员的青睐呢？](#1.1.2)
        * [1.1.3 历史](#1.1.3)
        * [1.1.4 Python的特点](#1.1.4)

    * [1.2 Python环境安装与Jupyter Notebook 使用](#1.2)
        * [1.2.1 环境安装](#1.2.1)
        * [1.2.2 Jupyter Notebook](#1.2.2)
        * [1.2.3 Latex](#1.2.3)
        
    * [1.3 语法](#1.3)
        * [1.3.1 行结构](#1.3.1)
        * [1.3.2 Empty lines](#1.3.2)
        * [1.3.3 Joining two lines](#1.3.3)
        * [1.3.4 Multiple statements on a Single line](#1.3.4)
        * [1.3.5 Indentation](#1.3.5)
        * [1.3.6 Style Guide](#1.3.6)
        * [1.3.7 Reserve Words](#1.3.7)
        * [1.3.8 build-in helper functions](#1.3.8)
        * [1.3.9 Zen](#1.3.9)
        
    * [1.4 Data Types and Structures 数据类型和数据结构](#1.4)        
        * [1.4.1 Data Types](#1.4.1)
        * [1.4.2 Collection types](#1.4.2)
        
    * [1.5 Conditions and Operators](#1.5)        
        * [1.5.1 Arithmetic Operations](#1.5.1)
        * [1.5.2 Boolean Operators or Comparison Operators](#1.5.2) 
        * [1.5.3 Assignment Operators](#1.5.3)
        * [1.5.4 Bitwise Operators](#1.5.4)
        * [1.5.5 Logical Operators](#1.5.5)
        * [1.5.6 Comparing null value](#1.5.6)
        * [1.5.7 Collection type operations](#1.5.7)
        * [1.5.8 三元操作符](#1.5.8)

    * [1.6 Conditional Statements](#1.6)        
        * [1.6.1 if-elif-else Statement](#1.6.1)
        * [1.6.1 try-except-raise Statements](#1.6.2)


***

# <font size=6><span id="1.1"> 1.1 Python 的前世今生</span></font>

## <font size=5><span id="1.1.1"> 1.1.1 Python的热度</span></font>

<img src="images/chapter01/Pythonredu.jpg">

### 再来一个榜单！
#### IEEE 的榜单结合 9 个数据来源的 11 个衡量指标（去年的排行榜有 12 个指标，今年少了 Dice job 网站信息，因其关闭了 API），权衡并发布了 47 种语言的排行榜，少于去年的 48 种。该排行榜允许读者根据自己的喜好或需求设定权重，如语言趋势或员工最关注的语言等。读者可以查看 Trending、Jobs、Open 等不同维度的编程语言排行数据。
<img src="images/chapter01/IEEErank.jpg">
##### https://spectrum.ieee.org/static/interactive-the-top-programming-languages-2018

#### Python 的排名从 2016 年开始就持续上升，去年顺利登顶，仅以 0.3 分的优势超越第二名 C 语言险夺第一。而今年，Python 依然高居榜首，且与第二名拉开差距。但是，今年的第二却不再是 C 语言，而是 C++ 了。C++ 的异军突起，使得 C 语言和 Java 都不得不「退位让贤」，分别降至第三和第四（去年它们分别位至第二和第三）。

## <font size=5><span id="1.1.2"> 1.1.2为什么 Python 会继续获得程序员的青睐呢？</span></font>
### 优点： 
### (1) python非常简单，非常适合人类阅读
### (2) 易学。python虽然是用c语言写的，但是它摈弃了c中非常复杂的指针，简化了python的语法。
### (3) Python是FLOSS（自由/开放源码软件）之一。简单地说，你可以自由地发布这个软件的拷贝、阅读它的源代码、对它做改动、把它的一部分用于新的自由软件中。Python希望看到一个更加优秀的人创造并经常改进。
### (4) 可移植性————由于它的开源本质，Python已经被移植在许多平台上
### (5) 在计算机内部，Python解释器把源代码转换成称为字节码的中间形式，然后再把它翻译成计算机使用的机器语言并运行。
### (6) Python既支持面向过程的函数编程也支持面向对象的抽象编程。在面向过程的语言中，程序是由过程或仅仅是可重用代码的函数构建起来的。在面向对象的语言中，程序是由数据和功能组合而成的对象构建起来的。与其他主要的语言如C++和Java相比，Python以一种非常强大又简单的方式实现面向对象编程。
### (7)  可扩展性和可嵌入性。如果你需要你的一段关键代码运行得更快或者希望某些算法不公开，你可以把你的部分程序用C或C++编写，然后在你的Python程序中使用它们。你可以把Python嵌入你的C/C++程序，从而向你的程序用户提供脚本功能。
### (8) 丰富的库。Python标准库确实很庞大。
### (9) Python确实是一种十分精彩又强大的语言。它合理地结合了高性能与使得编写程序简单有趣的特色。
### (10) 规范的代码。Python采用强制缩进的方式使得代码具有极佳的可读性。

## 缺点：
### (1) 运行速度，有速度要求的话，用C++改写关键部分吧。不过对于用户而言，机器上运行速度是可以忽略的。因为用户根本感觉不出来这种速度的差异。
### (2) 既是优点也是缺点，python的开源性是的Python语言不能加密，但是目前国内市场纯粹靠编写软件卖给客户的越来越少，网站和移动应用不需要给客户源代码，所以这个问题就是问题了。国随着时间的推移，很多国内软件公司，尤其是游戏公司，也开始规模使用他。
### (3) 构架选择太多（没有像C#这样的官方.net构架，也没有像ruby由于历史较短，构架开发的相对集中。Ruby on Rails 构架开发中小型web程序天下无敌）。不过这也从另一个侧面说明，python比较优秀，吸引的人才多，项目也多。

## <font size=5><span id="1.1.3"> 1.1.3 历史</span></font>

### Python的创始人为Guido van Rossum(吉多·范罗苏姆)。
<img src="https://gss3.bdstatic.com/7Po3dSag_xI4khGkpoWK1HF6hhy/baike/c0%3Dbaike272%2C5%2C5%2C272%2C90/sign=c0d38aab9d16fdfacc61cebcd5e6e731/d058ccbf6c81800a7ee4f88eb13533fa828b4728.jpg" width = 350>
### 1989年圣诞节期间，在阿姆斯特丹，Guido为了打发圣诞节的无趣，决心开发一个新的脚本解释程序，做为ABC语言的一种继承。之所以 选中Python（大蟒蛇的意思）作为程序的名字，是因为他是一个Monty 大蟒蛇飞行马戏团的爱好者。

### ABC是由Guido参加设计的一种教学语言（没听说过）。就Guido本人看来，ABC这种语言非常优美和强大，是专门为非专业程序员设计的。 但是ABC语言并没有成功，究其原因，Guido认为是非开放造成的。Guido决心在Python中避免这一错误（的确如此，Python与其它的语言 如C,C++和Java结合的非常好）。同时，他还想实现在ABC中闪现过但未曾实现的东西。



## <font size=5><span id="1.1.4"> 1.1.4 Python的特点</span></font>

### Python是一种脚本语言，它的语法表达优美易读。它具有很多优秀的脚本语言的特点：解释的，面向对象的，内建的高级数据结构，支持模块和包，支持多种平台，可扩展。而且它还支持交互式方式运行，图形方式运行。它的语法有很多与众不同的特性。
### 下面我分别说明一下：
### 1.1.4.1 运行方式
#### Python可以以命令行方式运行，也可以交互式方式运行，还具有图形集成环境，这样开发Python就相当方便。现在已经出现了许多用Python编写的可视化编程软件，用于实现象Delphi一样的功能。
### 1.1.4.2 面向对象
#### Python是一个真正的面向对象语言。它甚至支持异常的处理。如果学过Java，应该对这个不陌生。但其它的脚本语言，如PHP，好象就没有。这使得程序的编写更加清晰，而不需要许多的错误检查了。
### 1.1.4.3 模块和包
#### 这一点更象是Java。对于Java的支持，大家可以了解JPython。JPython是用Java写的Python，它完全支持Java，在这个环境下使用Python可以随意地使用Java的类库。
### 1.1.4.4 语言扩展
#### 可以用C、C++或Java为Python编写新的新言模块，如函数。或者与Python直接编译在一起，或者采用动态库装入方式实现。也专门有人编写了一个工具，可以实现为Python自动实现函数接口封装，这就是SWIG（Simplified Wrapper and Interface Generator)，或称做简单封装和接口生成器。
### 1.1.4.5 有趣的语法
#### Guido认为Python的语法是非常优美的。其中一点就是，块语句的表示不是C语言常用的{}对，或其它符号对，而是采用缩近表示法！有趣 吧。就这一点来说，Guido的解释是：首先，使用缩近表示法减少了视觉上的混乱，并且使程序变短，这样就减少了需要对基本代码单元注意的范围；其次，它 减少了程序员的自由度，更有利于统一风格，使得阅读别人的程序更容易。感觉还是不错的，就C语言来说，在if语句后面大括号的写法就好几种，不同的人喜欢 不同的样子，还不如统一起来，都不会看得别扭。
#### 在每个类或函数的定义后面，第一行可以是说明语句，根本不需要注释符标记。对于后面跟块语句的语句，后面应跟上一个冒号(:)。一行语句不能太长，因为没有行结束符，如果超长则要使用续行符(/)。还有一些有趣的比如说，象下面的一个比较处理，用C语言为：
#### if (2<a && a<5)
#### 用Python可以表示为
#### if (2<a<5) :
#### 当然不一定会有什么好处，但是它所宣称的优美的语言毕竟不是虚的。其它的大家可以通过学习自行了解。
### 1.1.4.6 Python可以做什么？
#### 如果在Windows下，如果想做一个图形界面程序，那么Python可以胜任。可以去wxPython.org看一下，它是一个Python的扩展库，可以用来实现跨平台的图形编程，它支持windows和unix/linux。在那个网 站上有一些演示，可以学习一下。Python可以做CGI。有些主页提供了对Python的CGI支持。现在还有一些用Python做的游戏。如果想了解 更多的Python可以做什么，可以去SourceForge上查找有关Python的项目，上面有很多东西。

[返回目录](#mulu)

___

# <font size=6><span id='1.2'> 1.2 Python环境安装与Jupyter Notebook 使用</span></font>

# 1.2.1 环境安装：https://www.anaconda.com/download/
<font size=5><span id="1.1.1"> 1.1.1 Python的热度</span></font>

<img src="images/chapter01/anaconda.jpg">
#                    Windows 系统     苹果mac操作系统   Linux操作系统



___

## <font size=5><span id="1.2.2"> 1.2.2 Jupyter Notebook </span></font>

## 1.2.2.1 什么是Jupyter Notebook

Jupyter Notebook（此前被称为 IPython notebook）是一个交互式笔记本，支持运行 40 多种编程语言。

Jupyter Notebook 的本质是一个 Web 应用程序，便于创建和共享文学化程序文档，支持实时代码，数学方程，可视化和 markdown。
用途包括：数据清理和转换，数值模拟，统计建模，机器学习等等

## 1.2.2.2 Notebook Name
The notebook name (filename) can be changed by double clicking on the name at the top  of the page, next to the jupyter logo.

可以在FILE下面改变文件名字,后缀ipynb

<img src="images\chapter01\notebook文件名.jpg" width="800" height="600">


## 1.2.2.3 This is a cell
Each cell could be changed to different supported format by selecting from the drop down list above. Or hit **m-key** for Markdown, **y-key** for Code and **r-key** to Raw.
### 用快捷键ESC+M 改变成markdown模式；ESC+Y 改变成code程序代码模式；用ESC+R改变成raw模式
This cell is of Markdown format

### markdown是啥?
Markdown是一种可以使用普通文本编辑器编写的标记语言，通过简单的标记语法，它可以使普通文本内容具有一定的格式。

Markdown的语法简洁明了、学习容易，而且功能比纯文本更强，因此有很多人用它写博客。

除此之外，现在由于我们有了RStudio这样的神级编辑器，我们还可以快速将Markdown转化为演讲PPT、Word产品文档、LaTex论文甚至是用非常少量的代码完成最小可用原型。在数据科学领域，Markdown已经广泛使用，极大地推进了动态可重复性研究的历史进程

In [ ]:
# This cell is of Code format

## Run a cell
To run a cell, just hit **shift-enter** or click on the **play** button in the toolbar or the **Cell:Run** menu item. 

To unrender the selected cell, double click on the cell.

In [4]:
x = 1
print(x)

1


When running a code cell, there will be a [ \* ] symbol next to the cell, when it is done, the sequence (order number) of the cell that was run will be displayed. 
## Output
Any output to console will be shown beneath the code cell

In [2]:
# helle world
print("Hello World!")

Hello World!


## Inspect Data
If the variable / data is on the last line of the cell, it will be printed as it as output.

In [ ]:
x

In [2]:
x = 2
type(x)

int

In [6]:
y = "String"
z = y + ' ' + str(x)
z

'String 1'

# Session
Each time a notebook is run, it is run on a new session. 

A session can be restarted by clicking on **Kernel:Restart** from the menu.

## Kernel
Jupyter notebook supports various kernels other than python and each cell's kernel may be changed be selecting the kernel from **Kernel:Change kernel:..**

The changed kernel is reflected on the right hand side of the menu bar.

# I LOVE YOU
## LOVE
###LOVE
####LOVE

***

## 1.2.2.4 Jupyter Notebook常用操作
### H：查看快捷键

## Esc：进入命令模式
### Enter：进入编辑模式
### M：markdown
### s：save保存
### x：删除
### a：above插入
### b：below插入
### Ctrl+Enter：运行cell
### shift+Enter：运行cell并进入下一个cell

### Alt+Enter : 运行本单元，在其下插入新单元

### X : 剪切选中的单元
### C : 复制选中的单元
### Shift-V : 粘贴到上方单元
### V : 粘贴到下方单元
### L：显示行号，在命令模式中的代码单元格按 `L` 打开数字。

### Esc + F   在代码中查找、替换，忽略输出。
### Esc + O   在 cell 和输出结果间切换。 (字母o)

### 选择多个 cell:  
#### Esc +Shift + J   或  Esc +Shift + Down   选择下一个 cell。
#### Esc +Shift + K   或  Esc + Shift + Up    选择上一个 cell。

#### 一旦选定 cell，可以批量删除 / 拷贝 / 剪切 / 粘贴 / 运行。当你需要移动 notebook 的一部分时这个很有用。

### Esc +Shift + M   合并 cell. 

In [3]:
i = 10
print(i)

j = 20

k = 30
print(i,j,k)

10
10 20 30


## <font size=5><span id="1.2.3"> 1.2.3 LaTex是什么？</span></font>

### LaTeX（LATEX是一种基于ΤΕΧ的排版系统，由美国计算机学家莱斯利·兰伯特（Leslie Lamport）在20世纪80年代初期开发，利用这种格式，即使使用者没有排版和程序设计的知识也可以充分发挥由TeX所提供的强大功能，能在几天，甚至几小时内生成很多具有书籍质量的印刷品。对于生成复杂表格和数学公式，这一点表现得尤为突出。因此它非常适用于生成高印刷质量的科技和数学类文档。这个系统同样适用于生成从简单的信件到完整书籍的所有其他种类的文档。

## 简单的规则：

### (1) 空格：LaTeX中空格用来隔开单词(英语一类字母文字)，多个空格等效于一个空格；对中文没有作用。 
### (2) 换行：用控制命令“\”,或“ \newline”. 
### (3) 分段：用控制命令“\par” 或空出一行。 
### (4) 换页：用控制命令“\newpage”或“\clearpage” 
### (5) 特殊控制字符：#，$, %, &, - ,{, }, ^, ~ 

### 要想输出这些控制符用下列命令：
\\# \\$ \\% \\& \\- \\{ \\} \\^{} \\~{} $\backslash$表示 "\".。


## 常用数学符号的 LaTeX 表示方法

### １、指数和下标可以用^和_后加相应字符来实现。比如：

#### 1     $a_{1}$   $x^{2}$  
#### 2     $e^{-\alpha t}$ 
#### 3     $a^{3}_{ij}$ 
#### 4     $e^{x^2} \neq {e^x}^2$

### 2、平方根（square root）的输入命令为：\sqrt，n 次方根相应地为:\sqrt[n]。方根符号的大小由 LATEX自动加以调整。也可用 \surd 仅给出符号,\quad表示退格，\qquad表示退两格。比如：

$$\sqrt{x}\qquad
\sqrt{ x^{2}+\sqrt{y} }\qquad \sqrt[3]{2}$$
$\surd[x^2 + y^2]$生成根号上没有横线的表示方式

### 3、命令\overline 和\underline 在表达式的上、下方画出水平线。比如：

$\overline{m+n} \qquad
\underline{m+n}$

#### 呈现效果： 
m+n¯¯¯¯¯¯¯¯¯¯¯¯¯m+n¯ \qquad 
m+n––––––m+n_ 
### 4、命令\overbrace 和\underbrace 在表达式的上、下方给出一水平的大括号。

$\underbrace{ a+b+\cdots+z }_{26}$

https://blog.csdn.net/testcs_dn/article/details/44229085

[返回目录](#mulu)

___

# <font size=6> <span id="1.3"> 1.3 语法</span></font>
## 语法解释器
## Python是一种脚本语言，即它是由解析器解释

## <font size=5><span id="1.3.1"> 1.3.1 行结构</span></font>
#### Python是基于行的语法形式
#### 当遇到换行符表示一句逻辑上结束了

In [ ]:
x = 1


## <font size=5><span id="1.3.2"> 1.3.2 空白行</span></font>
### Python解释器忽略那种只有一些空格、tabs，注释的一行，均认为是空白行


In [ ]:
# comment
    
    # more comments
	
        

### 注释从哈希字符#开始。所有的字符在行结束后都是注释的一部分，而Python解释器忽略了它们。
### 请注意，Python没有多行或块注释。

## <font size=5><span id="1.3.3"> 1.3.3 多行连接</span></font>

When a single line of codes is too long, it can be broken into two or more physical lines using the backslash (\) character.  
有时候遇到一行代码太长（打印和显示方便），会将一行用\符号隔开，表示下一行是接着的

In [4]:
variable_number_one   = 1
variable_number_two   = '2'
if variable_number_one == 1 \
    and variable_number_one < 2 \
    and variable_number_two == '2':
    print( "This is an example of line joining" )   # 用|表示行与行之间的连接

This is an example of line joining


In [12]:
a = 1
b = 2
if (a == 1 \
    and b ==2):
    print(a,b)

1 2


Line breaks are ignored inside pairs of [], {} or ().
行分割的做法，当遇到括弧的时候，就失效了

## <font size=5><span id="1.3.4"> 1.3.4 多个语句一行</span></font>

You can join two statements in a singale line by using a semicolon (;) to differentiate the two.
可以将多个语句写到一行，此时用分号隔开

In [13]:
str1 = "Statement 1"; str2 = "Statement 2"; print(str1); print(str2)

Statement 1
Statement 2


In [ ]:
# 等同于。。。。。。

## <font size=5><span id="1.3.5"> 1.3.5 缩进</span></font>

Whitespaces (spaces and tabs) are used to define blocks of codes. Where in other languages such as R or C/C++ curly braces ( { } ) are used.
用空格或者tab缩进一行代码；缩进的代码表示是一个块，从属于上面一句
### 严格的缩进是的python代码非常精简而且有层次
### 如果在正确的位置输入冒号，按下回车后，第二行代码会自动缩进

In [14]:
if x > 0:
    print("This statement")
    print("and this has the same indentation")

This statement
and this has the same indentation


The number of whitespaces is not fixed. But all statements whithin the block must be indented the same amount of spaces or tabls or spaces and tabs.  
缩进的空格数量不是固定的；但是在一个块内，要保持同样的缩进量

In [15]:
x = 5
if x > 0:
    print("x is greater than zero") #缩进了4个空格
    if x < 10:
     print ("x is single digit")  #缩进了1个空格
     print ("This statement has 4 spaces and a single tab indentation")

x is greater than zero
x is single digit
This statement has 4 spaces and a single tab indentation


In [ ]:
x = 5
if x > 0:
    print("x is greater than zero") #缩进了4个空格
    if x < 10:
        print ("x is single digit")  #缩进了1个空格
        print ("This statement has 4 spaces and a single tab indentation")

## <font size=5><span id="1.3.6"> 1.3.6 代码风格指南</span></font>
 * Use 4 spaces per indentation and no tabs.使用4个空格，不要用tabs
 * Do not mix tabs and spaces. Tabs create confusion and it is recommended to use only spaces. 不要混合tab和空格
 * Maximum line length : 79 characters which help users with a small display. 最多79个字符一行
 * Use blank lines to separate top-level function and class definitions and single blank line to separate methods definitions inside a class and larger blocks of code inside functions. 用空行区分不同程序块
 * When possible, put inline comments (should be complete sentences).  尽可能的注释
 * Use spaces around expressions and statements. 在表达式和语句中多用空格

In [ ]:
# 求方程根的代码
# 同学们现场编写一下

## <font size=5><span id="1.3.7"> 1.3.7 Reserve Words保留字</span></font>
The following identifiers are used as reserved words of the language, and cannot be used as ordinary identifiers.
一些关键字是系统自带的保留字，即不能用作为标识符

In [2]:
import keyword
keyword.kwlist

['False',
 'None',
 'True',
 'and',
 'as',
 'assert',
 'break',
 'class',
 'continue',
 'def',
 'del',
 'elif',
 'else',
 'except',
 'finally',
 'for',
 'from',
 'global',
 'if',
 'import',
 'in',
 'is',
 'lambda',
 'nonlocal',
 'not',
 'or',
 'pass',
 'raise',
 'return',
 'try',
 'while',
 'with',
 'yield']

In [ ]:
## <font size=5><span id="1.3.8"> 1.3.8 build-in functions内置函数</span></font>
### 如：print() input() 
### dir(\__builtins__)可以看到内置函数列表

In [3]:
dir(__builtins__)

['ArithmeticError',
 'AssertionError',
 'AttributeError',
 'BaseException',
 'BlockingIOError',
 'BrokenPipeError',
 'BufferError',
 'BytesWarning',
 'ChildProcessError',
 'ConnectionAbortedError',
 'ConnectionError',
 'ConnectionRefusedError',
 'ConnectionResetError',
 'DeprecationWarning',
 'EOFError',
 'Ellipsis',
 'EnvironmentError',
 'Exception',
 'False',
 'FileExistsError',
 'FileNotFoundError',
 'FloatingPointError',
 'FutureWarning',
 'GeneratorExit',
 'IOError',
 'ImportError',
 'ImportWarning',
 'IndentationError',
 'IndexError',
 'InterruptedError',
 'IsADirectoryError',
 'KeyError',
 'KeyboardInterrupt',
 'LookupError',
 'MemoryError',
 'ModuleNotFoundError',
 'NameError',
 'None',
 'NotADirectoryError',
 'NotImplemented',
 'NotImplementedError',
 'OSError',
 'OverflowError',
 'PendingDeprecationWarning',
 'PermissionError',
 'ProcessLookupError',
 'RecursionError',
 'ReferenceError',
 'ResourceWarning',
 'RuntimeError',
 'RuntimeWarning',
 'StopAsyncIteration',
 'StopIter

In [16]:
aa = 3
aaaa = 5
dir()  #python内置函数；不带参数时，返回当前范围内的变量、方法和定义的类型列表
        #带参数是，返回参数的属性、方法列表



['In',
 'Out',
 '_',
 '_2',
 '__',
 '___',
 '__builtin__',
 '__builtins__',
 '__doc__',
 '__loader__',
 '__name__',
 '__package__',
 '__spec__',
 '_dh',
 '_i',
 '_i1',
 '_i10',
 '_i11',
 '_i12',
 '_i13',
 '_i14',
 '_i15',
 '_i16',
 '_i2',
 '_i3',
 '_i4',
 '_i5',
 '_i6',
 '_i7',
 '_i8',
 '_i9',
 '_ih',
 '_ii',
 '_iii',
 '_oh',
 'a',
 'aa',
 'aaaa',
 'b',
 'exit',
 'get_ipython',
 'i',
 'j',
 'k',
 'quit',
 'str1',
 'str2',
 'variable_number_one',
 'variable_number_two',
 'x']

In [4]:
help(dir)


Help on built-in function dir in module builtins:

dir(...)
    dir([object]) -> list of strings
    
    If called without an argument, return the names in the current scope.
    Else, return an alphabetized list of names comprising (some of) the attributes
    of the given object, and of attributes reachable from it.
    If the object supplies a method named __dir__, it will be used; otherwise
    the default dir() logic is used and returns:
      for a module object: the module's attributes.
      for a class object:  its attributes, and recursively the attributes
        of its bases.
      for any other object: its attributes, its class's attributes, and
        recursively the attributes of its class's base classes.



In [17]:
help(print)   # 内置函数，查看函数或模块用途的详细说明

Help on built-in function print in module builtins:

print(...)
    print(value, ..., sep=' ', end='\n', file=sys.stdout, flush=False)
    
    Prints the values to a stream, or to sys.stdout by default.
    Optional keyword arguments:
    file:  a file-like object (stream); defaults to the current sys.stdout.
    sep:   string inserted between values, default a space.
    end:   string appended after the last value, default a newline.
    flush: whether to forcibly flush the stream.



Hit the **STOP** square button in the button ribbons/bar on top to continue to next cell.

In [29]:
#help(str)
import copy as cy
help(copy.copy)
a = [1, 2, 3]
b = cy.copy(a) // 浅拷贝

Help on function copy in module copy:

copy(x)
    Shallow copy operation on arbitrary Python objects.
    
    See the module's __doc__ string for more info.



In [22]:
b
id(b)

1373389870600

In [30]:
id(a)

1373388493320

In [25]:
a.append(7)

In [26]:
a

[1, 2, 3, 7]

In [27]:
b

[1, 2, 3]

## <font size=5><span id="1.3.9"> 1.3.9 The Zen of Python 禅</span></font>

In [ ]:
import this

#### References
https://www.w3resource.com/python/python-syntax.php

[返回目录](#mulu)

___

# <font size=6> <span id="1.4"> 1.4 数据类型和结构</span></font>

## <font size=5><span id="1.4.1"> 1.4.1 数据类型</span></font>


Everything in Python is an object.任何数据在python中都是一个对象  
Python objects do not need explicit declaration 最大的特点是，python对象不需要公开声明

### 整型

In [6]:
count =  5
type(count)

int

  
#### type() 函数如果你只有第一个参数则返回对象的类型

In [7]:
ii=500000000000000000000000000000000000000000000000000000000000000000
print(ii)
### python3已经对长整型无缝结合，即长度不受限制
### 如果说有限制，那也是计算机虚拟内存综述
### 所以python3可以进行大数计算

500000000000000000000000000000000000000000000000000000000000000000


In [32]:
sum = 0
for i in range(1,101):
    sum += i  # sum = sum + i
sum    
##此大数在其它语言是不太可能实现的！

5050

In [35]:
5 / 3

1.6666666666666667

In [9]:
round(5/3)  #四舍五入

2

In [10]:
round(5/3,3) #round函数的用法

1.667

In [16]:
5 // 2

2

In [19]:
round(5/3,1)

1.7

In [22]:
9 % 10

9

### 浮点型
#### python区分浮点型与整型的区别，就是有没有小数点

In [23]:
meters = 1.2
type(meters)

float

In [28]:
a = 1.2e12  # 1.2E12  E,e both are ok
print(a)

1200000000000.0


In [25]:
a = 12e12
type(a)

float

In [36]:


a= pi
print(a)

3.141592653589793


In [33]:
math.e

2.718281828459045

### 布尔类型

In [37]:
WWCodeRocks = True
type(WWCodeRocks)

bool

In [38]:
ww = true

NameError: name 'true' is not defined

In [36]:
True + True

2

In [43]:
eval(30-5)

TypeError: eval() arg 1 must be a string, bytes or code object

In [40]:
True * False

0

In [41]:
True / False

ZeroDivisionError: division by zero

#### True 相当于整型值1  False 0

### 复数

In [42]:
pi = 3.14J
print(pi)
type(pi)

3.14j


complex

In [43]:
a = 3 + 4j
type(a)

complex

In [44]:
b = 5 + 6J
print(a + b)

(8+10j)


## 变量
### 当把一个值赋给一个名字的时候，它会存储在内存之中，把这块内存称之为“变量variable”
### Python有些特殊，它并不是把值存储在变量之中，更像把名字贴在“值”的上面。

In [45]:
student = "张三"
id(student)

1798634607872

In [54]:
student = "李四"
id(student)

1798634532424

In [55]:
x = 5
id(x)

1862954176

In [44]:
x = 8
id(x)

1649372448

In [46]:
y = 8
id(y)

1649372448

In [45]:
id(8)

1649372448

## 关于变量的注意事项
### * 使用变量前，需要对其进行赋值
### * 变量名可以使用字母、数字、下划线命名，不能以数字开头（跟C语言 一样）
### * 大小写区分，即：Abc  abc 不同
### * 等于号=是赋值的意思，左边是名字，右边是值
### * 变量的命名虽说可以任意，还是要有范
#### 错误的名字：举例说明
#### 正确的方式：Num  \__abc__ _2b


### None Type
The None type represents a null object.  
None is the default return value of a Python function

In [58]:
ret = print('None returned')
type(ret)

None returned


NoneType

## Change data type 数据类型转换

In [47]:
X = int('ABCD', 16)
print(X)
type(X)

43981


int

In [60]:
a = '520'
b = int(a)
a, b
# c = 5.99
# d = int(c)
# c, d

('520', 520)

In [61]:
help(int)

Help on class int in module builtins:

class int(object)
 |  int(x=0) -> integer
 |  int(x, base=10) -> integer
 |  
 |  Convert a number or string to an integer, or return 0 if no arguments
 |  are given.  If x is a number, return x.__int__().  For floating point
 |  numbers, this truncates towards zero.
 |  
 |  If x is not a number or if base is given, then x must be a string,
 |  bytes, or bytearray instance representing an integer literal in the
 |  given base.  The literal can be preceded by '+' or '-' and be surrounded
 |  by whitespace.  The base defaults to 10.  Valid bases are 0 and 2-36.
 |  Base 0 means to interpret the base from the string as an integer literal.
 |  >>> int('0b100', base=0)
 |  4
 |  
 |  Methods defined here:
 |  
 |  __abs__(self, /)
 |      abs(self)
 |  
 |  __add__(self, value, /)
 |      Return self+value.
 |  
 |  __and__(self, value, /)
 |      Return self&value.
 |  
 |  __bool__(self, /)
 |      self != 0
 |  
 |  __ceil__(...)
 |      Ceiling of

In [62]:
int('0b100', base=0)

4

In [ ]:
float() #转成小数

In [48]:
a = '520'
b = float(a)
a, b

('520', 520.0)

In [63]:
c = 520
d = float(c)
c, d

(520, 520.0)

In [ ]:
str()  #将一个数或其他类型的书转换成字符串

In [64]:
a = 5.99
b = str(a)
b

'5.99'

In [65]:
c = str(5e15)
c

'5000000000000000.0'

## Multiple Assignment 同时赋值

Python allows assigning a single value to several variables simultaneously.

In [67]:
I = luv = U = False
print("All three variables: I luv U are")
print(I and luv and U)

All three variables: I luv U are
False


In [68]:
b, f, country = True, 2.0, "Singapore"
print(str(b) + ' ' + country)
f

True Singapore


2.0

## Delete assignment清除变量
As Python allows manual 'garbage collection' to delete a variable

In [69]:
del f
print(f)

NameError: name 'f' is not defined

[返回目录](#mulu)

## <font size=5><span id="1.4.2"> 1.4.2 Collection types -- Collection类型</span></font>
There's also a standard library module called collections that provides more specialized collection types other than the built-in types below  
collection这里指：各种数据的集合

## Strings  字符串
Strings are a continuous set of characters represented by quotation marks.

In [70]:
name = "Women Who Code"
type(name)

str

In [ ]:
'5'

In [76]:
a_string = 'Hello' + ' ' + "Women Who Code!"
print(a_string)
print("str[0]  :" + a_string[0])
print("str[2:5:2]:" + a_string[2:5:2]) # Python speak: slicing
print("str[2:] :" + a_string[2:])
print('''Triple quotes can\
span multiple lines''')

Hello Women Who Code!
str[0]  :H
str[2:5:2]:lo
str[2:] :llo Women Who Code!
Triple quotes canspan multiple lines


In [77]:
'Let's go'

SyntaxError: invalid syntax (<ipython-input-77-e2906f04b32e>, line 1)

In [78]:
'Let\'s go'

"Let's go"

In [79]:
print('Let\'s go')

Let's go


## 字符串可以看做引号内的一切东西，也叫做文本，与数字截然不同

In [ ]:
5 + 8

In [80]:
#str类型可以进行加法运算
'5'+'8'

'58'

In [49]:
# str 乘法
'I love you' * 3

'I love youI love youI love you'

In [87]:
bb = '1234ADSFDSAadf'
cc = bb[4:12]
print(cc)

ADSFDSAa


## 单引号、双引号，Python不挑剔，但必须成对出现

In [ ]:
stringa = "I Love Python"
print(stringa)

In [ ]:
stringa = "I Love Python'
print(stringa)

## 如果希望文本里面有单引号或者双引号怎么办？

In [ ]:
'Let's go'

In [50]:
'Let\'s go'   #用转义字符\进行转义

"Let's go"

In [51]:
"Let's go"   #用双引号来也可以

"Let's go"

## 转义字符

In [106]:
string = 'C\\'  #打印试试看
string

'C\\'

In [107]:
print(string)   #看看打印的效果

C\


In [90]:
'c:\now'

'c:\now'

In [6]:
print(r'c:\now')  # \n转义字符

c:\now


### 说明原因在于 \n 构成了回车换行

| 转义字符    | 描述                                         |
| ----------- | -------------------------------------------- |
| \(在行尾时) | 续行符                                       |
| \\          | 反斜杠符号                                   |
| \'          | 单引号                                       |
| \"          | 双引号                                       |
| \a          | 响铃                                         |
| \b          | 退格(Backspace)                              |
| \e          | 转义                                         |
| \000        | 空                                           |
| \n          | 换行                                         |
| \v          | 纵向制表符                                   |
| \t          | 横向制表符                                   |
| \r          | 回车                                         |
| \f          | 换页                                         |
| \oyy        | 八进制数，yy代表的字符，例如：\o12代表换行   |
| \xyy        | 十六进制数，yy代表的字符，例如：\x0a代表换行 |
| \other      | 其它的字符以普通格式输出                     |

In [93]:
string = 'C:\\now'
string

'C:\\now'

In [92]:
print(string)

C:
ow


In [52]:
string='D:\MARUJIN\18-TEACH-PYTHON\sound'  #不是我们想要的
print(string)

D:\MARUJIN8-TEACH-PYTHON\sound


In [53]:
string='D:\\MARUJIN\\18-TEACH-PYTHON\\sound'  #心太累
print(string)

D:\MARUJIN\18-TEACH-PYTHON\sound


In [22]:
string=r'D:\MARUJIN\18-TEACH-PYTHON\sound'  #加一个英文字母r
string

'D:\\MARUJIN\\18-TEACH-PYTHON\\sound'

In [15]:
print(string)

D:\MARUJIN\18-TEACH-PYTHON\sound


### 无论是否是原始字符串，都不能以反斜杠作为结尾
### 反斜杠置于文末，表示该字符串还没结束

In [16]:
string = 'I Love\'

SyntaxError: EOL while scanning string literal (<ipython-input-16-fa8ff4f56444>, line 1)

In [20]:
string = 'I Love\\'
print(string)

I Love\


### 思考：非要在文末加一个反斜杠，怎么实现？

In [99]:
pass

## 长字符串

卜算子·我住长江头

宋代：李之仪

我住长江头，  
君住长江尾。  
日日思君不见君，  
共饮长江水。 
  
此水几时休，  
此恨何时已。  
只愿君心似我心，  
定不负相思意。

### 打印试试

In [100]:
print("卜算子·我住长江头\n\n宋代：李之仪\n\n我住长江头，\n君住长江尾。\n日日思君不见君，\n共饮长江水。\n\n此水几时休，\n此恨何时已。\n只愿君心似我心，\n定不负相思意")

卜算子·我住长江头

宋代：李之仪

我住长江头，
君住长江尾。
日日思君不见君，
共饮长江水。

此水几时休，
此恨何时已。
只愿君心似我心，
定不负相思意


### 行数太多，太麻烦
### Python有办法

### 长字符串

### 用三重引号"""内容"""

In [54]:
aa = """曾经有一份真诚的爱摆在我的面前，
但是我没有珍惜，
等到失去的时候才后悔莫及，
尘世间最痛苦的事莫过于此。

如果上天可以给我个机会再来一次的话，
我会对这个女孩说我爱她，
如果非要在这份爱加上一个期限，
我希望是一万年！"""
print(aa)

曾经有一份真诚的爱摆在我的面前，
但是我没有珍惜，
等到失去的时候才后悔莫及，
尘世间最痛苦的事莫过于此。

如果上天可以给我个机会再来一次的话，
我会对这个女孩说我爱她，
如果非要在这份爱加上一个期限，
我希望是一万年！


In [101]:
print(aa)

NameError: name 'aa' is not defined

In [103]:
bb = '''   悯农
锄禾日当午
汗滴禾下土
谁知盘中餐
粒粒皆辛苦'''
print(bb)

   悯农
锄禾日当午
汗滴禾下土
谁知盘中餐
粒粒皆辛苦


[返回目录](#mulu)

## Tuples 元组
Tuples are collection of arbitrary objects. Equivalent to records or structs in other languages.  
Tuples are enclosed in parentheses ( ( ) ) and cannot be updated.
* 元组用圆括弧构成的一些数据的集合
* 元组里面的元素不可以修改

In [55]:
t1 = ( "WWCode", 100000 , 0.5 ) # org name, members, proportion of engineers
t2 = 'Singapore', 1160.5
t_singleton = ('We',) # singleton
t_empty = ()
print(type(t1)); print(type(t2))
print(t_singleton);
type(t_empty)

<class 'tuple'>
<class 'tuple'>
('We',)


tuple

In [57]:
t3 = 3, 5  # 自动组成了一个元组
t3

(3, 5)

Except for the case of the empty tuple, it is really the commas, not the parentheses, that tell Python it is a tuple.

In [62]:
t = ('a',)
type(t)

tuple

In [70]:
t4 = (1,)
t5 = [1,]
print(type(t4))
print(type(t5))

<class 'tuple'>
<class 'list'>


## Mutable object types: 可变对象类型
Some collection classes are mutable. The methods that add, subtract, or rearrange their members in place, and don’t return a specific item, never return the collection instance itself but **None**.
* 可变表示这个对象可以增加，删除，重新排列  

## Lists 列表  
* Lists in Python contains items separated by commans and enclosed within square brackets, '[' ']'.   
* 列表用方括弧来表示
* Lists are highly flexible, and are like mutable tuples.  
* 列表是可变的，就好像可变的元组
* Values stored in a list can be accessed using the slice operator ([] and [:])
* 元组可以用下标方式获取

In [108]:
list = [ "WWCode", 786 , 2.23, 'singapore', 70.2 ]
print(list[0:3])
type(list)

['WWCode', 786, 2.23]


list

In [109]:
list[0][2:]

'Code'

### List comprehension
List comprehension or Listcomp for short, is a powerful way of buildling lists.

In [112]:
symbols = '$¢£¥€¤'
codes = [ord(symbol) for symbol in symbols]
codes

list1 = [i for i in range(10)]
list1.sort(reverse = True)
print(list1)
list2 = sorted(list1)
print(list2)
list3 = list(filter(lambda x: x % 2 == 0,dict(zip(list2, list1)))) # 列表，元组，字典，集合都是可迭代的对象
tup1 = tuple(zip(list2, list1))
list4 = sorted(tup1, key = lambda x:x[1] ) # 对元组才可以进行如此类型的排序 
# 多级排序
print(list3)
help(filter)
set(list2)
dict(zip(list2, list1))
list4
   # 对元组多级排序：
   # list_sorted_first = sorted(tuple_list, key=lambda x: (dict_type_1[x[1]], dict_type_1[x[2]], dict_type_1[x[3]]))


[9, 8, 7, 6, 5, 4, 3, 2, 1, 0]
[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
[0, 2, 4, 6, 8]
Help on class filter in module builtins:

class filter(object)
 |  filter(function or None, iterable) --> filter object
 |  
 |  Return an iterator yielding those items of iterable for which function(item)
 |  is true. If function is None, return the items that are true.
 |  
 |  Methods defined here:
 |  
 |  __getattribute__(self, name, /)
 |      Return getattr(self, name).
 |  
 |  __iter__(self, /)
 |      Implement iter(self).
 |  
 |  __new__(*args, **kwargs) from builtins.type
 |      Create and return a new object.  See help(type) for accurate signature.
 |  
 |  __next__(self, /)
 |      Implement next(self).
 |  
 |  __reduce__(...)
 |      Return state information for pickling.



[(9, 0),
 (8, 1),
 (7, 2),
 (6, 3),
 (5, 4),
 (4, 5),
 (3, 6),
 (2, 7),
 (1, 8),
 (0, 9)]

In [114]:
help(ord) #返回符号的unicode编码

Help on built-in function ord in module builtins:

ord(c, /)
    Return the Unicode code point for a one-character string.



## dictionary 字典
Python's dictionaries are hash tables, and consists of key-value pairs.  
Dictionaries are enclosed by curly braces '{' '}' and values can be assigned and accessed like a dictionary/hash table.  
Order of the elements is not maintained in a dictionary. 

In [114]:
dict1 = {'name':'Women Who Code Singapore', 
        'org':'WWCode', 
        'city':'Singapore',
        'members':1260}
print(dict1['org'])
type(dict1)

WWCode


dict

In [116]:
dict2 = {'赵':1, 
        '钱':2,
        '孙':3,
        '李':4,
        '周':5,
        '吴':6,
        '郑':7,
        '王':8,
        '冯':9,
        '陈':10,
        '褚':11,
        '卫':12,
        }

In [117]:
print(dict2)

{'赵': 1, '钱': 2, '孙': 3, '李': 4, '周': 5, '吴': 6, '郑': 7, '王': 8, '冯': 9, '陈': 10, '褚': 11, '卫': 12}


In [118]:
dict2['王']

8

## Sets 集合  
* Sets are a methematical concept, they are a lot like dictionaries with keys but no corresponding values. 
* 跟数学的概念很相似，类似于字典的键，但没有对应的值
* Sets are enclosed by curly braces, elements seperated by comma, '{','}'.  
* 用花括弧
* Sets do not support indexing or slicing, and do not have inherent order.  
* 不支持下标应用和切片

In [119]:
wwcode_asia_networks = {'Bangalore','Beijing','Chennai','Delhi','Gujarat','Hong Kong','Kuala Lumpur','Manila','Pune','Rajasthan','Shanghai','Singapore','Taipei','Tel-Aviv','Tokyo'}
type(wwcode_asia_networks)

set

In [120]:
print(wwcode_asia_networks)

{'Tokyo', 'Bangalore', 'Kuala Lumpur', 'Rajasthan', 'Tel-Aviv', 'Singapore', 'Beijing', 'Hong Kong', 'Manila', 'Gujarat', 'Delhi', 'Pune', 'Shanghai', 'Taipei', 'Chennai'}


In [122]:
wwcode_asia_networks[1]

TypeError: 'set' object does not support indexing

In [129]:
a = {2}
type(a)

set

#### Reference
https://www.tutorialspoint.com/python/python_variable_types.htm

[返回目录](#mulu)

# <font size=6> <span id="1.5"> 1.5 Conditions and Operators</span></font>

## <font size=5><span id="1.5.1"> 1.5.1 算术运算符</span></font>


In [121]:
5//3

1

运算符|	描述|	实例
-| :-: | -
=	|简单的赋值运算符	|c = a + b 将 a + b 的运算结果赋值为 c
+=	|加法赋值运算符	|c += a 等效于 c = c + a
-=	|减法赋值运算符	|c -= a 等效于 c = c - a
*=	|乘法赋值运算符	|c *= a 等效于 c = c * a
/=	|除法赋值运算符	|c /= a 等效于 c = c / a
%=	|取模赋值运算符	|c %= a 等效于 c = c % a
**=	|幂赋值运算符	|c \*\*= a 等效于 c = c \*\* a
//=	|取整除赋值运算符	|c //= a 等效于 c = c // a

### 优先级
#### 先乘除，后加减，有括号先计算括号


In [130]:
-3 * 2 + 5 / -2 -4

-12.5

## <font size=5><span id="1.5.2"> 1.5.2 布尔运算和比较运算</span></font>


In [131]:
print(a == b) # equals
print(a != b) # not equals
print(a > b)  # greater than
print(a < b)  # lesser than
print(a >= b) # greater than or equal
print(a <= b) # lesser than or equal

False
True


TypeError: '>' not supported between instances of 'set' and 'float'

运算符|描述|实例
:- | :-: | -: 
    ==      |等于 - 比较对象是否相等|    10 == 20 输出结果 ：False 
||[1, 2, 3] == [1, 2, 3] 输出结果：True
!=|	不等于 - 比较两个对象是否不相等	|“abc” != “abc” 输出结果：Flase 
||[1, 2, 3] != [1, 3, 2] 输出结果：True
>	|大于 - 返回x是否大于y	|5 > 8 输出结果：Flase
<	|小于 - 返回x是否小于y	|4 < 9 输出结果：True
>=|	大于等于 - 返回x是否大于等于y	|5 >= 3 输出结果：True
<=	|小于等于 - 返回x是否小于等于y	|7 <= 4 输出结果：Flase

## <font size=5><span id="1.5.2"> 1.5.3 Assignment Operators赋值运算符</span></font>


In [132]:
i = 100;  print(i)
i += 2; print(i)
i -= 3; print(i)
i *= 4; print(i)
i /= 5; print(i)
i %= 6; print(i)
i **= 7; print(i)
i //= 2; print(i)

100
102
99
396
79.2
1.2000000000000028
3.5831808000000596
1.0


## <font size=5><span id="1.5.4"> 1.5.4 Bitwise Operators 位运算符 </span></font>


In [25]:
j = 60; k = 13; 
print("j = 00{0:b} k = 00{0:b}".format(j,k))
print("j = 0000{0:b} ".format(j & k) + str(j & k)) # AND
print("j = 00{0:b} ".format(j | k) + str(j | k))   # OR
print("j = 00{0:b} ".format(j ^ k) + str(j ^ k))   # XOR
print("j =  {0:b} ".format( ~j ) + str( ~j ))      # Ones complement, i.e. flipping every bit
print("j = {0:b} ".format(j<<2) + str( j<<2 ))     # Left Shift
print("j = 0000{0:b} ".format(j>>2) + str( j>>2 )) # Right Shift

j = 00111100 k = 00111100
j = 00001100 12
j = 00111101 61
j = 00110001 49
j =  -111101 -61
j = 11110000 240
j = 00001111 15


| 运算符 | 描述                                                         | 实例                                                         |
| ------ | ------------------------------------------------------------ | ------------------------------------------------------------ |
| & | 按位与运算符：参与运算的两个值,如果两个相应位都为1,则该位的结果为1,否则为0 | (a & b) 输出结果 12 ，二进制解释： 0000 1100     |
|｜|      按位或运算符：只要对应的二个二进位有一个为1时，结果位就为1。 | (a｜b) 输出结果 61 ，二进制解释： 0011 1101         |
| ^      | 按位异或运算符：当两对应的二进位相异时，结果为1              | (a ^ b) 输出结果 49 ，二进制解释： 0011 0001 |
| ~      | 按位取反运算符：对数据的每个二进制位取反,即把1变为0,把0变为1 。~x 类似于 -x-1 | (~a ) 输出结果 -61 ，二进制解释： 1100 0011，在一个有符号二进制数的补码形式。 |
| <<     | 左移动运算符：运算数的各二进位全部左移若干位，由 << 右边的数字指定了移动的位数，高位丢弃，低位补0。 | a << 2 输出结果 240 ，二进制解释： 1111 0000                 |
| >>     | 右移动运算符：把">>"左边的运算数的各二进位全部右移若干位，>> 右边的数字指定了移动的位数 | a >> 2 输出结果 15 ，二进制解释： 0000 1111                  |

## <font size=5><span id="1.5.5"> 1.5.5 Logical Operators 逻辑运算符</span></font>
## and or not

In [133]:
not True

False

In [134]:
not 0

True

In [135]:
not 4

False

In [136]:
3 < 4 < 5

True

In [140]:
3 < 4 < 2

False

In [141]:
3 < 4 and 4 < 5

True

In [142]:
3 < 5 or 8 > 5

True

In [123]:
a_string = "Women Who Code"
print("Women" in a_string)
print("Men" not in a_string)
print(len("Women Who Code") is len(a_string))
print(len("Hello World!") is not len(a_string))

True
True
True
True


In [51]:
LeapYear=int(input('输入年份（四位数）'))
#LeapYear = 2100

if ((LeapYear % 400 == 0) or ((LeapYear % 4 ==0) and (LeapYear %100 != 0))):
    print(LeapYear,"is a LeapYear")
else:
    print(LeapYear,"is not a LeapYear")



输入年份（四位数）2005
2005 is not a LeapYear


In [144]:
a = input("请输入一个数")
type(a)

请输入一个数"22"


str


## is 和  is not 运算符
## 与==以及!=的区别

### is 用于判断两个变量引用对象是否为同一个， == 用于判断引用变量的值是否相等。

In [146]:
x = 5
y = 5
print(x == y)
print(x is y)
print(id(x))
print(id(y))

True
True
1649372352
1649372352


In [148]:
help(id)

Help on built-in function id in module builtins:

id(obj, /)
    Return the identity of an object.
    
    This is guaranteed to be unique among simultaneously existing objects.
    (CPython uses the object's memory address.)



In [149]:
id(1)

1649372224

In [150]:
id("abc")

1373320727440

In [151]:
id([1, 2, 3])

1373390189640

In [153]:
x, y = "abcabcabcabcabcabcabcabcabcabc", "abcabcabcabcabcabcabcabcabcabc"
print(x == y)
print(x is y)
print(id(x))
print(id(y))

True
True
1373390278128
1373390278128


In [43]:
# 数组比较
x = [1, 2, 3]
y = [1, 2, 3]
print(x == y)
print(x is y)    # 内存空间不等
print(id(x))
print(id(y))

True
False
2405124970824
2405124915400


In [155]:
# 元组比较
x = (1, 2, 3)
y = (1, 2, 3)
print(x == y)
print(x is y)
print(id(x))
print(id(y))

True
False
1373390376896
1373390355336


In [156]:
# 字典比较
x = {"id": 1, "name": "Tom", "age": 18}
y = {"id": 1, "name": "Tom", "age": 18}
print(x == y)
print(x is y)# 判断引用的对象的内存地址是否相同
print(id(x))
print(id(y))

True
False
1373390541232
1373390541160


In [157]:
# 集合比较
x = set([1, 2, 3])
y = set([1, 2, 3])
print(x == y)
print(x is y)
print(id(x))
print(id(y))

True
False
1373388574056
1373374483816


In [135]:
# 赋值后比较
x = [1, 2, 3]
y = x
print(x == y)
print(x is y)
print(id(x))
print(id(y))

True
True
1798634778696
1798634778696


## <font size=5><span id="1.5.6"> 1.5.6 Comparing null value 空值比较</span></font>


In [140]:
none_type = None
print(none_type is None)

## <font size=5><span id="1.5.7"> 1.5.7 Collection type operations</span></font>

In [141]:
'Slicing a string'[6:11]

'g a s'

In [142]:
# Slice position 3 to 9 every 2 step
'0123456789'[3:9:2]

'357'

In [46]:
# accessing items from end of collection
'0123456789'[-1]

'9'

In [143]:
# Most collection types could be sorted
sorted('987654321')

['1', '2', '3', '4', '5', '6', '7', '8', '9']

## - 动手试试看（切片试验）

In [ ]:
# try


In [144]:
# any() returns True if any member of the collection is true
# 判断是否为空对象
any([0,1,2])

True

In [158]:
# all() returns True if all the members are true
all([0,1,2]) #所有元素不为0、''、False或者x为空对象，则返回True，否则返回False

False

### String specific operations

In [169]:
print('Concat' + 'enation')
print('Multiplication ' * 2)
print('upper'.upper())
print('Lower'.lower())
print('capitalise'.capitalize())
print('center'.center(48))
print('ljust'.ljust(48))
print('rjust'.rjust(48))
print('find i'.find('i')) # return the lowest index where the given substring is found, -1 on failure
print('index of i'.index('i')) # raises a ValueError exception on failure
print('rfind i'.rfind('i')) # return the highest index of the given substring
print(' join '.join(['list','of','strings']))
print('split this string on passed    space separator'.split())
print("""split
lines""".splitlines())
print('partition'.partition('i')) # returns the first string, the separator, and the remaining string
print('   strip white    spaces   '.strip()) # default is whitespace, or strip passed characters
print('    left strip    '.lstrip())
print('    right strip    '.rstrip())
print('replace a with X'.replace('a','X'))
a = 2
print('Formats the string, x={}'.format(a))

Concatenation
Multiplication Multiplication 
UPPER
lower
Capitalise
                     center                     
ljust                                           
                                           rjust
1
0
6
list join of join strings
['split', 'this', 'string', 'on', 'passed', 'space', 'separator']
['split', 'lines']
('part', 'i', 'tion')
strip white    spaces
left strip    
    right strip
replXce X with X
Formats the string, x=2


In [ ]:
# try

### Boolean string operations 字符串判断函数

In [145]:
print('startswith'.startswith('start')) 

True


In [ ]:
print('endswith'.endswith('endswith'))

In [171]:
print('isalpha'.isalpha())  #是否有字母组成

True


In [172]:
print('isdigit'.isdigit()) # 是否由数值组成

False


In [173]:
print('isalnum'.isalnum())  #是否由字母和数字组成

True


### Dictionary Operations 字典判断函数

In [174]:
wwcodesg = {'name':'Women Who Code Singapore', 
                 'org':'WWCode', 
                 'city':'Singapore',
                 'members':1260}

In [175]:
print(wwcodesg.keys())
print(wwcodesg.values())
print(wwcodesg.items())

dict_keys(['name', 'org', 'city', 'members'])
dict_values(['Women Who Code Singapore', 'WWCode', 'Singapore', 1260])
dict_items([('name', 'Women Who Code Singapore'), ('org', 'WWCode'), ('city', 'Singapore'), ('members', 1260)])


In [177]:
print(wwcodesg.get('name1'))  # 避免异常的出现

None


In [179]:
print(wwcodesg.items())

dict_items([('name', 'Women Who Code Singapore'), ('org', 'WWCode'), ('city', 'Singapore'), ('members', 1260)])


In [182]:
print(wwcodesg.pop('org'))

KeyError: 'org'

In [181]:
print(wwcodesg.items())

dict_items([('name', 'Women Who Code Singapore'), ('city', 'Singapore'), ('members', 1260)])


In [185]:
print(wwcodesg.setdefault('org1', 'Women Who Code1')) #字典 setdefault() 函数和get() 方法类似, 如果键不存在于字典中，将会添加键并将值设为默认值。

Women Who Code1


In [187]:
wwcodesg['a'] = 1   # 给字典添加

In [191]:
print(wwcodesg.items())
fistName = 3

dict_items([('name', 'Women Who Code Singapore'), ('city', 'Singapore'), ('members', 1260), ('org1', 'Women Who Code1'), ('a', 1)])


### Set operations 集合判断函数

In [192]:
set1 = {'1','2','3','4','5',}

In [193]:
print('4' in set1)

True


In [194]:
print({'1','2','4'}.issubset(set1))

True


In [195]:
print({'2','3','4'} <= set1 ) # subset or the same
print({'1','2','3','4','5'} < set1 ) # is subset

True
False


In [196]:
print(set1.issuperset({'2','3','4'}))

True


In [197]:
print(set1 >= {'2','3','4'})

True


In [200]:
print(set1 > {'4','5','6'})

False


In [207]:
print({'5','6'}.union(set1))
help(str)

{'4', '6', '3', '1', '5', '2'}
Help on class str in module builtins:

class str(object)
 |  str(object='') -> str
 |  str(bytes_or_buffer[, encoding[, errors]]) -> str
 |  
 |  Create a new string object from the given object. If encoding or
 |  errors is specified, then the object must expose a data buffer
 |  that will be decoded using the given encoding and error handler.
 |  Otherwise, returns the result of object.__str__() (if defined)
 |  or repr(object).
 |  encoding defaults to sys.getdefaultencoding().
 |  errors defaults to 'strict'.
 |  
 |  Methods defined here:
 |  
 |  __add__(self, value, /)
 |      Return self+value.
 |  
 |  __contains__(self, key, /)
 |      Return key in self.
 |  
 |  __eq__(self, value, /)
 |      Return self==value.
 |  
 |  __format__(...)
 |      S.__format__(format_spec) -> str
 |      
 |      Return a formatted version of S as described by format_spec.
 |  
 |  __ge__(self, value, /)
 |      Return self>=value.
 |  
 |  __getattribute__(self,

In [206]:
set2 = {'1', '7', '8', '9'} 
print(set1.intersection(set2)) #交集 

{'1'}


In [ ]:
print(set1)
print(set2)
set3 = set1 & set2 # same as intersection 交集
print(set3)
print(set2.difference(set1)) # difference()差 set2里面有，set1里面没有的

In [ ]:
print(set2 - set1) # same as difference

In [ ]:
print(set1)
print(set2)
print(set2.symmetric_difference(set1)) # 返回：合集-交集

In [ ]:
print(set1)
print(set2)
print(set1 ^ set2 )# 返回：合集-交集

### Set Modifier Operations 集合修改运算
Set modifiers update the set's value(s) in place

In [ ]:
set1 = {'1','2','3','4','5',}

In [208]:
set1.update({'9'}); print(set1)

{'3', '5', '2', '9', '1', '4'}


In [ ]:
set2 = {'6', '7', '8'}
set2 |= {'1'}; print(set2) # same as update

In [ ]:
set2.intersection_update(set1); print(set2)

In [ ]:
set2 &= set1; print(set2) # same as intersection update

In [ ]:
asean.difference_update(wwcode_asean); print(asean)
wwcode_asia -= asean; print(wwcode_asia) # same as difference update
asean.symmetric_difference_update(wwcode_asean); print(asean)
wwcode_asia ^= asean; print(wwcode_asia) # same as symmetric difference update
wwcode_asean.add('Jakarta'); print(wwcode_asean)
wwcode_asean.remove('Jakarta'); print(wwcode_asean)
wwcode_asean.discard('Jakarta'); print(wwcode_asean) # does not return KeyError if element not found
print(wwcode_asean.pop()) # removes and returns an arbitrary member from the set.
wwcode_asean.clear(); print(wwcode_asean) # removes all elements

## <font size=5><span id="1.5.8"> 1.5.8 三元操作符</span></font>

In [209]:
x = 5
y = 8
if x < y:
    small =  x
else:
    small = y
small

5

In [210]:
# rewrite
small = x if x < y else y
small

5

[返回目录](#mulu)

# <font size=6> <span id="1.6"> 1.6 Conditional Statements 条件判断语句</span></font>

## <font size=5><span id="1.6.1"> 1.6.1 if-elif-else Statement</span></font>


In [211]:
if(i==1):
    print("Happy")

Happy


In [212]:
if(i<1):
    print("Happy")
elif(i==1.0):
    print("Weekend")

Weekend


In [148]:
i=1
if(i<1):
    print("Happy")
    print("yes")
elif(i!=1.0):
    print("Weekend")
else:
    print("Ahead!")

Ahead!


In [ ]:
## 百分制转成 五分制
score = int(input('请输入成绩：'))

if 90 <= score <= 100:
    print('A')
if 80 <= score < 90:
    print('B')
if 70 <= score < 80:
    print('C')
if 60 <= score < 70:
    print('D')
if 0 <= score < 60:
    print('F')
if score < 0 or score > 100:
    print('input error')



## Play

In [213]:
"""玩一个游戏吧"""
temp=input("猜一个数字：")
guess=int(temp)
if guess == 8:
    print("你太厉害了，被你猜中了")
    print("不好意思，没奖品")
else:
    print("猜错啦，正确答案是8！")
print("Game over, Bye")

猜一个数字：153
猜错啦，正确答案是8！
Game over, Bye


## <font size=5><span id="1.6.2"> 1.6.2 try-except-raise Statements</span></font>

Try statements are useful when you know there are possibility of an error but you want to handle these errors and not stop the program.

In [215]:
while True:
    try:
        x = int(input("Please enter a number: "))
        break
    except ValueError:
        print("Oops!  That was not a valid number.  Try again...")

Please enter a number: abc
Oops!  That was not a valid number.  Try again...
Please enter a number: 1


### pass in try statements
Use pass to continue without performing an action.  
An except clause may name multiple exceptions as a parenthesized tuple.

In [149]:
try:
    print(Hello)
except (RuntimeError, TypeError, NameError):
    pass

### raise in try statements

In [216]:
try:
    raise NameError('HiThere')
except NameError:
    print('An exception flew by!')

An exception flew by!


In [217]:
try:
    raise NameError('HiThere')
except NameError:
    print('An exception flew by!')
    raise

An exception flew by!


NameError: HiThere

### else in try statements
The try … except statement has an optional else clause, which, when present, must follow all except clauses.  
It is useful for code that must be executed if the try clause does not raise an exception. 

In [218]:
arg = 'README.md'
try:
    f = open(arg, 'r')
except OSError:
    print('cannot open', arg)
else: # 如果try执行成功，则就执行else内的语句
    print(arg, 'has', len(f.readlines()), 'lines')
    f.close()

cannot open README.md


### finaly in try statements
A finally clause is always executed before leaving the try statement, whether an exception has occurred or not.

In [224]:
def divide(x, y):
    try:
        result = x / y
    except ZeroDivisionError:
        print("division by zero!")
    except:
        print("GGwp")
    else:
        print("result is", result)
    finally:
        print("executing finally clause")

In [220]:
divide(2, 0)

division by zero!
executing finally clause


In [221]:
divide(4,2)

result is 2.0
executing finally clause


In [225]:
divide("2", "1")

GGwp
executing finally clause


#### References
https://www.learnpython.org/en/Conditions  
https://www.tutorialspoint.com/python/python_basic_operators.htm  

[返回目录](#mulu)